crates/flipperzero/src/furi/kernel.rs

```diff
@@ -0,0 +1,89 @@
+//! Furi Kernel primitives.
+
+use flipperzero_sys as sys;
+use ufmt::derive::uDebug;
+
+use crate::furi;
+
+/// Check if CPU is in IRQ; or kernel running and IRQ is masked.
+pub fn is_irq_or_masked() -> bool {
+    unsafe { sys::furi_kernel_is_irq_or_masked() }
+}
+
+/// Check if kernel is running.
+pub fn is_running() -> bool {
+    unsafe { sys::furi_kernel_is_irq_or_masked() }
+}
+
+/// Kernel lock state.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, uDebug)]
+pub enum LockState {
+    /// normal scheduling
+    Unlocked = 0,
+    /// scheduling paused
+    Locked = 1,
+}
+
+impl From<i32> for LockState {
+    fn from(value: i32) -> Self {
+        match value {
+            0 => LockState::Unlocked,
+            1 => LockState::Locked,
+            x => panic!("invalid lock state value: {x}"),
```

Comment:
Panicking `From` is quite bad, since it can be called by any user. `TryFrom` with on-demand `expect` should be better.

Reply:
Looking at the implementation, I believe the intent is for the `i32` to act as either an error `Status` or a boolean lock state (0: unlocked, 1: locked). Effectively, it ends up like a cast to `bool` (e.g. `if (lock_state) { // LOCKED } else { // UNLOCK }`. Hard to be sure since no code in `flipper-firmware` actually uses the results of these functions...

I've modified the behaviour to treat `0` as Unlocked and non-`0` as Locked, so it won't panic on invalid values. This would also allow more specific Locked states to be added in the future.